### DISCUSSION TITLE
Vertical transform of cloud and precipitation

### DISCUSSION THREAD

**Discussion by marstonsward** [xgcm/xgcm#590] [Q&A]
Trying to understand the docs for conservative transformation. Is this the correct method for cloud and precipitation variables?
If my new vertical coordinate looks like: `np.arange(-0.5, 80.5, 0.5)`  what would the cell bounds look like (`np.arange(-0.5, 80.5, 0.5) + 0.5`)? It is a bit confusing from the example which gives `zc = np.arange(1.0,12) + 0.5` and `theta = np.linspace(0, 3, 20)`. 
And if I understand the examples, the transformed products will be on the edges? Just curious how to get back to a center value?

**Comment by jbusecke** [comment]
Hi @Marston, the conservative interpolation is meant to be used on extensive variables (e.g. some transport integrated over cell depth). 
I am not sure what variables you are working with here. Could you provide some detail on what kind of variables you have and **on which grid position they are located**?

I can try to outline the method like this: 

Goal: Get some variable `data` (on the center position) into a new vertical coordinate based on an array 'target_coord'

For this we need the values of `target_coord` on the location of the cell bounds surrounding `data`. Often this is not the case for numerical model output (since `target_coord` in reality is often a tracer which is located on a cell center). The example mostly deals with creating an `outer` grid position (`zc`) and then interpolating the temperature data (`theta`) onto that outer position to satisfy this requirement. 

Note: While the resulting data is not located on the outer bounds, I see that the naming in the example is chosen poorly. The name of the new dimension (here 'theta_outer') has nothing to do with the grid position, but instead is just based on the name of the intput. 
I could have done something like: 
```python
phi_transformed_cons = grid.transform(ds.phi,
                                      'Z',
                                      theta_target,
                                      method='conservative',
                                      target_data=ds.some_name)
```
and the resulting dimension would be named `some_name`. I think we should change this in the docs for clarity.

**Reply by marstonsward** [reply]
Thanks for the outline. XGCM is new and this API takes a bit of getting used to. I have successfully tested the linear transform for a UM model output for temperature. This parameter was defined in the vertical on pressure levels, but I needed it to be defined on MSL. Since I have the GMH data on pressure levels, it is an easy transformation. What XGCM gives me a fast means of doing this transformation in one fell swoop. Now I would like to do the same type of transformation but for 3D cloud cover and specific humidity, which as not something I want to do a linear transformation on. Therefore, I was looking at a conservative method. Maybe something such NN would also be great.
Hope this made my use case a bit clearer.

**Reply by jbusecke** [reply]
Unfortunately I am not sure about most of these abbreviations 😜. I think I need a bit of vocabulary here for UM, GMH, MSL, and NN.

If your use case is to preserve the vertical integral of a tracer concentration (like specific humidity is?), you can multiply the tracer concentration with the vertical grid distance (`tracer_extensive=tracer*dz`), conservatively transform both the vertical distance and the `tracer_extensive` variables and then divide the values again to get the concentration in the new coordinate system.

**Reply by marstonsward** [reply]
Apologies: 
Unified Model
Geometric Height
Height above mean sea level (MSL). I should have written height msl
Nearest Neighbour

I'm not sure SH (specific humidity) is a tracer as it can reach very high levels in the tropics and near the surface. I have to look up that definition. But clouds are a good example as they occur in discrete intervals.
